lib/awful/keygrabber.lua

```diff
@@ -45,7 +207,411 @@ function keygrabber.stop(g)
+--
+-- @DOC_text_awful_keygrabber_release_key_EXAMPLE@
+--
+-- Please note that modkeys are accepted as release keys as long as the
+-- keyboard layouts use the IBM PC 104-key naming convention for the key names.
+--
+-- @property release_key
+-- @param[opt=nil] string|table release_key
+-- @see release_event
+
+--- The event on which the keygrabbing will be terminated.
+--
+-- the valid values are:
+--
+-- * "press" (default)
+-- * "release"
+--
+-- @property release_event
+-- @param string
+-- @see release_key
+
+--- Whether or not to execute the key press/release callbacks when keybindings are called.
+--
+-- When this property is set to false, the `keyreleased_callback` and
+-- `keypressed_callback` callbacks will be executed even when the event was
+-- consumed by a `keybinding`.
+--
+-- By default, keybindings block those callbacks from being executed.
+--
+-- @property mask_event_callback
+-- @param[opt=true] boolean
+-- @see keybindings
+-- @see keyreleased_callback
+-- @see keypressed_callback
+
+--- Do not call the callbacks on modifier keys (like `Control` or `Mod4`) events.
+-- @property mask_modkeys
+-- @param[opt=false] boolean
+-- @see mask_event_callback
+
+--- Export all keygrabber keybindings as root (global) keybindings.
+--
+-- When this is enabled, calling all of the keygrabber object `keybinding`s will
+-- will create root `awful.key` and will automatically starts the grabbing.
+--
+-- Use this with caution. If many keygrabber or "real" root keybindings are set
+-- on the same key combination, they are all executed and there is almost no
+-- safe way to undo that. Also, while `is_enabled` can be used to temporarely
+-- turn off this behavior, it currently cannot be detached completely. Make sure
+-- the `keygrabber` that use this option have a single instance.
+--
+-- @property export_keybindings
+-- @param[opt=false] boolean
+
+--- The root (global) keybinding to start this keygrabber.
+--
+-- Instead of adding an entry to `root.keys` or `rc.lua` `globalkeys` section,
+-- this property can be used to take care of everything. This way, it becomes
+-- easier to package modules using keygrabbers.
+--
+-- @DOC_text_awful_keygrabber_root_keybindings_EXAMPLE@
+--
+-- @property root_keybindings
+-- @see export_keybindings
+-- @see keybindings
+
+--- The keybindings associated with this keygrabber.
+--
+-- The keybindings syntax is the same as `awful.key` or `awful.prompt.hooks`. It
+-- consists of a table with 4 entries.
+--
+-- * `mods` A table with modifier keys, such as `shift`, `mod4`, `mod1` (alt) or
+--  `control`.
+-- * `key` The key name, such as `left` or `f`
+-- * `callback` A function that will be called when the key combination is
+--  pressed.
+-- * `description` A table various metadata to be used for `awful.hotkeys_popup`.
+--
+-- @property keybindings
+-- @param table
+-- @see export_keybindings
+-- @see root_keybindings
+
+--- Define a list of keys that are allowed instead of using `release_key`.
```

Comment:
I fail to understand what this means (and I only noticed that because the code implementing this confused me).

Does this mean that when any key that is not allowed is pressed, the keygrabber is automatically stopped? I think that should be spelt out more explicitly.
`If any key is pressed that is not in this list, the keygrabber is stopped`